### PR TITLE
Cache probe column when key has many entries

### LINKED ISSUE #8295: Optimize join probe stage if one hash table key has many entries
## Enhancement
In TiFlash, the entry of a join hash table is <key, RowRefList>, where `RowRefList` is a list of the rows that has the same key, and each element if the `RowRefList` is a pair of <block_index, row_index>. During hash join probe stage, if a probe row has an existing key in the hash table, it will combine all the matched rows in the `RowRefList` using [`AddFound`](https://github.com/pingcap/tiflash/blob/e597b78a6f5595ba50ffe839daca798ced1d1bc4/dbms/src/Interpreters/JoinPartition.cpp#L1296), inside `AddFound`, it just iterator over the `RowRefList` and insert the data row by row
https://github.com/pingcap/tiflash/blob/e597b78a6f5595ba50ffe839daca798ced1d1bc4/dbms/src/Interpreters/JoinPartition.cpp#L1321-L1329

If the list is very long (~10000), it will take a lot if time in `AddFound`. We need to find a way to optimize this implementation.

A straightforward optimization is to cache `AddFound` result of such key, so next time when a probe row which has the same key comes, we can just use `insertRange` to batch insert the data instead of iterator over `RowRefList` and insert the data row by row

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #8295

Problem Summary:

### What is changed and how it works?
1. add `list_length` and `cache_column_info` in RefList of Join hash table
2. cache and reuse cached columns if a join key has many duplicated entries
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```
mysql> desc t;
+---------+---------+------+------+---------+-------+
| Field   | Type    | Null | Key  | Default | Extra |
+---------+---------+------+------+---------+-------+
| from_id | int(11) | YES  |      | NULL    |       |
| to_id   | int(11) | YES  |      | NULL    |       |
| value   | int(11) | YES  |      | NULL    |       |
+---------+---------+------+------+---------+-------+
mysql> select count(*) from t;
+----------+
| count(*) |
+----------+
|   393216 |
+----------+
mysql> select count(*), from_id, to_id, value from t group by from_id, to_id, value;
+----------+---------+-------+-------+
| count(*) | from_id | to_id | value |
+----------+---------+-------+-------+
|   131072 |       2 |     3 |     2 |
|   131072 |       3 |     4 |     3 |
|   131072 |       1 |     2 |     1 |
+----------+---------+-------+-------+
```
Before this pr
```
mysql>  select count(*) from t t1 join t t2 on t1.from_id = t2.to_id and t1.value > t2.value;
+-------------+
| count(*)    |
+-------------+
| 34359738368 |
+-------------+
1 row in set (3 min 21.22 sec)
```
After this pr
```
mysql>  select count(*) from t t1 join t t2 on t1.from_id = t2.to_id and t1.value > t2.value;
+-------------+
| count(*)    |
+-------------+
| 34359738368 |
+-------------+
1 row in set (1 min 16.07 sec)
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/JoinPartition.cpp [hunk 1]
```diff
@@ -1239,6 +1282,65 @@ struct Adder<ASTTableJoin::Kind::LeftOuterSemi, ASTTableJoin::Strictness::Any, M
...
+        {
+            added_columns[j]->insertRangeFrom(*cached_column_info->columns[j], 0, rows_added);
+        }
+    }
+}
+
+Columns constructCachedColumns(MutableColumns & added_columns, size_t rows, size_t columns)
+{
+    Columns cached_columns;
+    size_t start_offset = added_columns[0]->size() - rows;
```
Comment (yibin87) [inline dbms/src/Interpreters/JoinPartition.cpp:1330]: Better check size >= rows here.

File: dbms/src/Interpreters/JoinHashMap.h [hunk 2]
```diff
@@ -33,15 +33,29 @@ struct RowRef
     {}
 };
 
+struct CachedColumnInfo
+{
+    std::mutex mu;
+    Columns columns;
+    bool generate_cached_columns = false;
```
Comment (yibin87) [inline dbms/src/Interpreters/JoinHashMap.h:40]: An enum value instead of boolean value may make it more clean.

Comment (windtalker) [reply dbms/src/Interpreters/JoinHashMap.h]: OK, add an enum here.

File: dbms/src/Interpreters/JoinPartition.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Interpreters/JoinPartition.cpp]: OK, add an assert

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/Interpreters/JoinPartition.cpp [hunk 3]
```diff
@@ -1239,6 +1294,80 @@ struct Adder<ASTTableJoin::Kind::LeftOuterSemi, ASTTableJoin::Strictness::Any, M
...
+        {
+            for (size_t j = 0; j < columns; ++j)
+                cached_columns.push_back(added_columns[j]->cut(start_offset, rows));
+        }
+    }
+    std::unique_lock lock(cached_column_info->mu);
+    if (!cached_columns.empty())
+    {
+        assert(cached_column_info->columns.empty());
+        cached_column_info->columns.insert(
```
[the remark this answers is not in the document]
Comment (windtalker) [reply dbms/src/Interpreters/JoinPartition.cpp]: Done

**Pull request merged**